<a href="https://cognitiveclass.ai"><img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width = 400> </a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>


## Capstone Project for Introduction to Deep Learning with Keras


In this project, you will build a regression model using the Keras library to model the same data about concrete compressive strength that we used in labs 3. For your convenience, the data can be found here again: https://cocl.us/concrete_data. To recap, the predictors in the data of concrete strength include:

Cement
Blast Furnace Slag
Fly Ash
Water
Superplasticizer
Coarse Aggregate
Fine Aggregate



The four parts of the capstone:
<h5> A. Build a baseline model</h5>
<h5> B. Normalize the data </h5>
<h5> C. Increate the number of epochs </h5>  
<h5> D. Increase the number of hidden layers </h5>  



<h2>Regression Models with Keras</h2>

<h3>Capstone Project<h3>    
<h5> 1. Download and Clean dataset</h5>
<h5> 2. Build a Regression Model with Keras </h5>
<h5> 3. Train and Test the Network. </h5>    


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#item31">Download and Clean Dataset</a>  
2. <a href="#item32">Import Keras</a>  
3. <a href="#item33">Build a Neural Network</a>  
4. <a href="#item34">Train and Test the Network</a>  

</font>
</div>


<a id="item31"></a>


## Download and Clean Dataset


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [1]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [2]:
# Download Data
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 


#### Let's check how many data points we have.


In [3]:
# Checking size of data
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
# Checking for Null values
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # input data or predictors drawn form data
target = concrete_data['Strength'] # set target to strenght

<a id="item2"></a>


Let's do a quick sanity check of the predictors and the target dataframes.


In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.


In [9]:
# Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.
# Normalize the data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to _n_cols_ since we will need this number when building our network.


In [10]:
n_cols = predictors_norm.shape[1] # number of predictors
print(n_cols)

8


<a id="item1"></a>


<a id='item32'></a>


## Import Keras


Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.


#### Let's go ahead and import the Keras library


In [11]:
import keras

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

As you can see, the TensorFlow backend was used to install the Keras library.


Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.


In [12]:
from keras.models import Sequential
from keras.layers import Dense

## 1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn


In [13]:
from sklearn.model_selection import train_test_split
predictors_train, predictors_test, target_train, target_test = train_test_split( predictors, target, test_size=0.3, random_state=4)
print ('Train set:', predictors_train.shape,  target_train.shape)
print ('Test set:', predictors_test.shape,  target_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


<a id='item33'></a>


## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

One hidden layer of 10 nodes, and a ReLU activation function

In [14]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [15]:
%whos

Variable                Type         Data/Info
----------------------------------------------
Dense                   type         <class 'keras.layers.core.Dense'>
Sequential              type         <class 'keras.models.Sequential'>
concrete_data           DataFrame          Cement  Blast Furna<...>\n[1030 rows x 9 columns]
concrete_data_columns   Index        Index(['Cement', 'Blast F<...>],\n      dtype='object')
keras                   module       <module 'keras' from '/ho<...>kages/keras/__init__.py'>
n_cols                  int          8
np                      module       <module 'numpy' from '/ho<...>kages/numpy/__init__.py'>
pd                      module       <module 'pandas' from '/h<...>ages/pandas/__init__.py'>
predictors              DataFrame          Cement  Blast Furna<...>\n[1030 rows x 8 columns]
predictors_norm         DataFrame            Cement  Blast Fur<...>\n[1030 rows x 8 columns]
predictors_test         DataFrame         Cement  Blast Furnac<...>n\n[309

<a id="item4"></a>


<a id='item34'></a>


## 2. Train and Test the Network


Let's call the function now to create our model.


In [16]:
# build the model
model = regression_model()

Train the model on the training data using 50 epochs.

In [17]:
# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 0s - loss: 1718.7386 - val_loss: 1234.4781
Epoch 2/50
 - 0s - loss: 1697.8396 - val_loss: 1222.4203
Epoch 3/50
 - 0s - loss: 1677.4982 - val_loss: 1210.4945
Epoch 4/50
 - 0s - loss: 1658.3636 - val_loss: 1199.0019
Epoch 5/50
 - 0s - loss: 1639.4429 - val_loss: 1187.8925
Epoch 6/50
 - 0s - loss: 1621.0532 - val_loss: 1177.1395
Epoch 7/50
 - 0s - loss: 1602.6225 - val_loss: 1166.6485
Epoch 8/50
 - 0s - loss: 1584.4741 - val_loss: 1155.9676
Epoch 9/50
 - 0s - loss: 1566.1436 - val_loss: 1145.7885
Epoch 10/50
 - 0s - loss: 1547.6347 - val_loss: 1135.3694
Epoch 11/50
 - 0s - loss: 1529.1951 - val_loss: 1124.8791
Epoch 12/50
 - 0s - loss: 1509.8490 - val_loss: 1114.7038
Epoch 13/50
 - 0s - loss: 1490.4928 - val_loss: 1104.2443
Epoch 14/50
 - 0s - loss: 1470.4921 - val_loss: 1093.4732
Epoch 15/50
 - 0s - loss: 1449.8407 - val_loss: 1082.6660
Epoch 16/50
 - 0s - loss: 1428.6221 - val_loss: 1071.5075
Epoch 17/50
 - 0s - loss: 1406.1964

In [18]:
# Evaluate the model on the test data using `evaluate`
print("Evaluate on test data")
results = model.evaluate(predictors_test, target_test, batch_size=128)
print("test loss, test acc:", results)

# Generate predictions (probabilities -- the output of the last layer)
# on new data using `predict`
print("Generate predictions for 3 samples")
predictions = model.predict(predictors_test[:3])
print("predictions shape:", predictions.shape)

Evaluate on test data
309/309 [==============================] - 0s 8us/step
test loss, test acc: 27837482.498381875
Generate predictions for 3 samples
predictions shape: (3, 1)


## 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [19]:
from sklearn.metrics import mean_squared_error
times = 50
l = []

for time in range(times):
    print("Number of time: ", time+1)
    predictors_train, predictors_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=4)
    model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)
    
    pred = model.predict(predictors_test)

    mse = mean_squared_error(target_test, pred)
    l.append(mse)
    print(" ")

Number of time:  1
Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 0s - loss: 512.0032 - val_loss: 496.1683
Epoch 2/50
 - 0s - loss: 493.7938 - val_loss: 481.3905
Epoch 3/50
 - 0s - loss: 476.9191 - val_loss: 466.5339
Epoch 4/50
 - 0s - loss: 460.3803 - val_loss: 452.7655
Epoch 5/50
 - 0s - loss: 444.9666 - val_loss: 438.7505
Epoch 6/50
 - 0s - loss: 430.1395 - val_loss: 425.9691
Epoch 7/50
 - 0s - loss: 416.3211 - val_loss: 412.8560
Epoch 8/50
 - 0s - loss: 402.9021 - val_loss: 400.6098
Epoch 9/50
 - 0s - loss: 390.3854 - val_loss: 389.1664
Epoch 10/50
 - 0s - loss: 378.4726 - val_loss: 377.8435
Epoch 11/50
 - 0s - loss: 367.2075 - val_loss: 367.2868
Epoch 12/50
 - 0s - loss: 356.6379 - val_loss: 356.7791
Epoch 13/50
 - 0s - loss: 346.5921 - val_loss: 347.1792
Epoch 14/50
 - 0s - loss: 337.1841 - val_loss: 337.9410
Epoch 15/50
 - 0s - loss: 328.2209 - val_loss: 328.7312
Epoch 16/50
 - 0s - loss: 319.6911 - val_loss: 320.5124
Epoch 17/50
 - 0s - loss: 311.5802 - val_loss: 3

## 5. Report the mean and the standard deviation of the mean squared errors.

In [20]:
mean = np.mean(l)
std = np.std(l)

print("Mean: ", mean)
print("Standard Deviation: ", std)

Mean:  48.038129698529474
Standard Deviation:  24.3328285883534


This notebook is part of a course on **Coursera** called _Introduction to Deep Learning & Neural Networks with Keras_
